plugins/cookieauth.js

```diff
@@ -13,167 +13,230 @@
 // limitations under the License.
 'use strict';
 
-// this the the 'cookieauth' request handler.
-// Instead of passing the authentication credentials using HTTP Basic Auth with every request
-// we exchange the credentials for a cookie which we remember and pass back with each
-// subsequent request.
-var async = require('async');
-var debug = require('debug')('cloudant:cookieauth');
-var stream = require('stream');
-var u = require('url');
-var nullcallback = function() {};
-
-module.exports = function(options) {
-  var requestDefaults = options.requestDefaults || {};
-  var request = require('request').defaults(requestDefaults);
-  var jar = request.jar();
-  var cookieRefresh = null;
-
-  // make a request using cookie authentication
-  // 1) if we have a cookie or have no credentials, just try the request
-  // 2) otherwise, get session cookie
-  // 3) then try the request
-  var cookieRequest = function(req, callback) {
-    // deal with absence of callback
-    if (typeof callback !== 'function') {
-      callback = nullcallback;
+const debug = require('debug')('cloudant:plugins:cookieauth');
+const lockFile = require('lockfile');
+const request = require('request');
+const tmp = require('tmp');
+const u = require('url');
+
+const BasePlugin = require('./base.js');
+
+function noop() {}
```

Comment:
As per #217 perhaps this noop could be removed and the lock functions extracted to somewhere common. Happy for that change to happen after merge of this and #217 if it makes things easier.